### PR TITLE
Making EnvelopeEncryptionFilterIT a template test so it can test HashiCorp Vault too

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Enhancement / new feature

### Description

_Please describe your pull request_

### Additional Context

_Why are you making this pull request?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonar-Lint warnings are addressed or are justifiably ignored.
- [ ] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

**Comment by k-wall** [comment]
@robobario @gracegrimwood @SamBarker this is still a work in progress, but I'd value some feedback.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+import io.kroxylicious.kms.provider.hashicorp.vault.VaultKmsService;
+import io.kroxylicious.kms.provider.hashicorp.vault.VaultResponse;
+import io.kroxylicious.proxy.config.FilterDefinition;
+import io.kroxylicious.proxy.config.FilterDefinitionBuilder;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class VaultTemplateInvocationContext implements TestTemplateInvocationContext {
+    private static final String VAULT_TOKEN = "rootToken";
+    private static final String HASHICORP_VAULT = "hashicorp/vault:1.15";
```
Comment (SamBarker) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java:43]: does this need to be kept in sync with the version defined in the pom?

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+                            @Override
+                            public TestKekManager getTestKekManager() {
+                                return new TestKekManager() {
+                                    @Override
+                                    public CompletionStage<Void> generateKek(String alias) {
+                                        Objects.requireNonNull(alias);
+
+                                        try {
+                                            create(alias);
```
Comment (SamBarker) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java:80]: technically we could inline the require non null here 
```suggestion
                                            create(Objects.requireNonNull(alias));
```

I'm on the fence if that improves readability or not.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                                        return runVaultCommand(new TypeReference<>() {
+                                        }, "vault", "write", "-f", "transit/keys/%s/rotate".formatted(keyId));
+                                    }
+
+                                    private <D> D runVaultCommand(TypeReference<VaultResponse<D>> valueTypeRef, String... args) {
+                                        try {
+                                            var execResult = vaultContainer.execInContainer(args);
+                                            int exitCode = execResult.getExitCode();
+                                            assertThat(exitCode).isZero();
+                                            var response = new ObjectMapper().readValue(execResult.getStdout(), valueTypeRef);
```
Comment (SamBarker) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java:117]: nit picking but creating a new object mapper for every command feels expensive.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/InMemoryTemplateInvocationContext.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+                                        return kms.resolveAlias(alias)
+                                                .handle((u, t) -> {
+                                                    if (t instanceof UnknownAliasException) {
+                                                        var kekId = kms.generateKey();
+                                                        kms.createAlias(kekId, alias);
+                                                        return null;
+                                                    }
+                                                    else {
+                                                        throw new AlreadyExistsException("key with alias " + alias + "already exists");
```
Comment (robobario) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/InMemoryTemplateInvocationContext.java:76]:
```suggestion
                                                        throw new AlreadyExistsException("key with alias " + alias + " already exists");
```

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/InMemoryTemplateInvocationContext.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+                        };
+                    }
+                },
+                (BeforeTestExecutionCallback) extensionContext -> {
+                    kms = IntegrationTestingKmsService.newInstance().buildKms(new IntegrationTestingKmsService.Config(kmsId.toString()));
+                },
+                (AfterTestExecutionCallback) extensionContext -> {
+                    IntegrationTestingKmsService.delete(kmsId.toString());
+                });
+    }
```
Comment (robobario) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/InMemoryTemplateInvocationContext.java:103]: generally feels like a big telescoping set of braces to read, maybe worth extracting the TestContext resolver into a method or class to so it's easier to see the list of `Extensions`

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                                            Thread.currentThread().interrupt();
+                                            throw new RuntimeException(e);
+                                        }
+                                    }
+
+                                };
+                            }
+                        };
+                    }
+                },
```
Comment (robobario) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java:133]: as in the other comment, I think we should extract the TestContext resolver to make the extension list easier to read

**Review by robobario — APPROVED** [review]
LGTM with nits

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java (hunk 1 above)
Comment (k-wall) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java:43]: There is a default in org.testcontainers.vault.VaultContainer#DEFAULT_TAG but unfortunately it is really old (1.1.3) so I think we need our own constant.  I do want to the constant to come from a single location.

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/TemplateKekSelector.java [hunk 7]
```diff
@@ -57,6 +58,9 @@ public CompletionStage<Map<String, K>> selectKek(@NonNull Set<String> topicNames
                     .map(
                             topicName -> kms.resolveAlias(evaluateTemplate(topicName))
                                     .exceptionallyCompose(e -> {
+                                        if (e instanceof CompletionException && e.getCause() != null) {
+                                            e = e.getCause();
+                                        }
                                         if (e instanceof UnknownAliasException) {
                                             return CompletableFuture.completedFuture(null);
                                         }
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/TemplateKekSelector.java]: Good point, I special cased both possibilities and added supporting unit tests.

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/TemplateKekSelector.java [hunk 8]
```diff
@@ -57,6 +58,9 @@ public CompletionStage<Map<String, K>> selectKek(@NonNull Set<String> topicNames
                     .map(
                             topicName -> kms.resolveAlias(evaluateTemplate(topicName))
                                     .exceptionallyCompose(e -> {
+                                        if (e instanceof CompletionException && e.getCause() != null) {
+                                            e = e.getCause();
```
[the remark this answers is not in the document]
Comment (k-wall) [reply kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/TemplateKekSelector.java]: Specifying an API with methods that return completionstage is tricky.

1. throw new RuntimeException("bang");
1. return CompletableFuture.failedStage(new RuntimeException("bang!");
1. return CompletaleFuture.completedStage(null).thenApply(v -> throw new RuntimeException("bang!"));

how does one javadoc that? `@throws` covers only the first case. should your `@return` say something like 'hey expect the possibility that the completionstage might be completed exceptionally with anything you see in the `@throws` cause.. and by the way.. it just might be wrapped in a `CompletionException` too.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java (hunk 2 above)
Comment (k-wall) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java:80]: When I use Java `Objects.requireNotNull` I'm using them as you would use a precondition clause (in languages such as Eiffel or Ada where such things are first class).  Given those influences, it feels strange to bleed the preconditions (state invariant) into what the method actually does.

That all said, I'm happy to leave adopt a different practice if that's what we all want.

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/TemplateKekSelector.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (SamBarker) [reply kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/TemplateKekSelector.java]: Sorry I just meant re-assigning `e` I agree with your analysis otherwise.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java (hunk 2 above)
Comment (SamBarker) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java:80]: I just find it strange to read a call I know returns and we "ignore" the return value. 

So I lean slightly towards using it inline, but happy to agree any convention and stick to it.

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/TemplateKekSelector.java [hunk 9]
```diff
@@ -58,15 +58,11 @@ public CompletionStage<Map<String, K>> selectKek(@NonNull Set<String> topicNames
                     .map(
                             topicName -> kms.resolveAlias(evaluateTemplate(topicName))
                                     .exceptionallyCompose(e -> {
-                                        if (e instanceof CompletionException && e.getCause() != null) {
-                                            e = e.getCause();
-                                        }
-                                        if (e instanceof UnknownAliasException) {
+                                        if (e instanceof UnknownAliasException
+                                                || (e instanceof CompletionException ce && ce.getCause() instanceof UnknownAliasException)) {
```
Comment (SamBarker) [inline kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/TemplateKekSelector.java:62]: In a true break from tradition for me, I wonder if (in some other PR) we should introduce a util method to check if a future exception is an X. 
For example this would become 
```suggestion
if (ExceptionUtils.is(e, UnknownAliasException.class)) {
```

and exception utils looks something like 

```java
public boolean is(Exception e, Class<?> targetType) {
    return targetType.isAssignableFrom(e.getClass()) || (CompletionException.class.isAssignableFrom(e) && targetType.isAssignableFrom(e.getCause().getClass())
}
```
Note this has not met a compiler...

**Review by SamBarker — APPROVED** [review]

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java (hunk 2 above)
Comment (k-wall) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java]: @robobario do you have any preference?

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/InMemoryTemplateInvocationContext.java (hunk 5 above)
Comment (k-wall) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/InMemoryTemplateInvocationContext.java]: It got refactored several times, please take another look.

File: kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/TemplateKekSelector.java (hunk 9 above)
Comment (k-wall) [reply kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/TemplateKekSelector.java:62]: yes, that sounds like a good idea.  I expect we'll meet the same pattern over and over.

**Comment by k-wall** [comment]
@SamBarker @robobario Could you take another look?  I'm hoping that this is approaching landable and we can cut the next release very soon.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/EnvelopeEncryptionFilterIT.java [hunk 10]
```diff
@@ -218,8 +221,12 @@ void produceAndConsumeEncryptedAndPlainTopicsAtSameTime(KafkaCluster cluster, To
         }
     }
 
+    // TODO express this test as a unit test and consider doing away with the test as the IT level.
     @TestTemplate
     void shouldGenerateOneDek(KafkaCluster cluster, Topic topic, TestKmsFacade testKmsFacade) throws Exception {
+        assumeThatCode(testKmsFacade::getKms).doesNotThrowAnyException();
```
Comment (k-wall) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/EnvelopeEncryptionFilterIT.java:230]: I'd really prefer that getKms wasn't part of the API at all.  The ITs shouldn't need to care.  I'd like to revisit this, but not block getting this merged and 0.4.0 out.

File: kroxylicious-kms-test-support/src/main/java/io/kroxylicious/kms/service/TestKmsFacade.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+    /**
+     * Gets the configuration Kroxylicious will need to use to connect to the underlying KMS.
+     * @return service configuration.
+     */
+    C getKmsServiceConfig();
+
+    /**
+     * Returns the actual {@link Kms} in-use by Kroxylicious.  This is an optional method.
+     * It is expected most implementation will throw an {@link UnsupportedOperationException}.
```
Comment (k-wall) [inline kroxylicious-kms-test-support/src/main/java/io/kroxylicious/kms/service/TestKmsFacade.java:56]:
```suggestion
     * It is expected most implementations will throw an {@link UnsupportedOperationException}.
```

File: kroxylicious-kms-provider-hashicorp-vault-test-support/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsFacade.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+                }
+                catch (IOException e) {
+                    throw new UncheckedIOException("Failed to run vault command: %s".formatted(Arrays.stream(args).toList()), e);
+                }
+                catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(e);
+                }
+            }
+
```
Comment (k-wall) [inline kroxylicious-kms-provider-hashicorp-vault-test-support/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsFacade.java:119]:
```suggestion
```

File: kroxylicious-kms-provider-hashicorp-vault-test-support/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsFacade.java [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+                .withVaultToken(VAULT_TOKEN)
+                .withEnv("VAULT_FORMAT", "json")
+                .withInitCommand(
+                        "secrets enable transit");
+        vaultContainer.start();
+    }
+
+    @Override
+    public void stop() {
+
```
Comment (k-wall) [inline kroxylicious-kms-provider-hashicorp-vault-test-support/src/main/java/io/kroxylicious/kms/provider/hashicorp/vault/VaultKmsFacade.java:55]:
```suggestion
```

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/EnvelopeEncryptionFilterIT.java [hunk 14]
```diff
@@ -338,28 +257,19 @@ void shouldGenerateOneDek(KafkaCluster cluster) throws Exception {
...
-     * exploratory test examining what the client will see/do when decryption fails - looking to verify
-     * - behaviour is reasonable
-     * - the user has a chance to understand what's wrong.
-     * behaviour when KMS throws a KmsException a) during produce, b) during fetch
-     */
-    private FilterDefinition encryptionFilterDefinition(UUID kmsId) {
-        return new FilterDefinitionBuilder("EnvelopeEncryption")
-                .withConfig("kms", IntegrationTestingKmsService.class.getSimpleName())
-                .withConfig("kmsConfig", Map.of("name", kmsId.toString()))
+    private FilterDefinition getEncryptionFilterDefinition(TestKmsFacade<?, ?, ?> testKmsFacade) {
```
Comment (SamBarker) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/EnvelopeEncryptionFilterIT.java:266]: personal preference nit:

```suggestion
    private FilterDefinition buildEncryptionFilterDefinition(TestKmsFacade<?, ?, ?> testKmsFacade) {
```

get implies accessing something that exists already and thus shared but `build` implies that its being constructed for the caller.

File: kroxylicious-kms-provider-kroxylicious-inmemory-test-support/src/test/java/io/kroxylicious/kms/provider/kroxylicious/inmemory/InMemoryTestKmsFacadeTest.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+class InMemoryTestKmsFacadeTest extends AbstractTestKmsFacadeTest<Config, UUID, InMemoryEdek> {
+
+    InMemoryTestKmsFacadeTest() {
+        super(new InMemoryTestKmsFacadeFactory());
+    }
+
+    @Test
+    void classAndConfig() {
+        try (var facade = factory.build()) {
+            facade.start();
```
Comment (SamBarker) [inline kroxylicious-kms-provider-kroxylicious-inmemory-test-support/src/test/java/io/kroxylicious/kms/provider/kroxylicious/inmemory/InMemoryTestKmsFacadeTest.java:27]: Should we also assert that facade is not null? Or is the test failing with a Null pointer exception good enough? Probably.

File: kroxylicious-kms-provider-kroxylicious-inmemory-test-support/pom.xml [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    <dependencies>
+        <!-- project dependencies - runtime and compile -->
+        <dependency>
+            <groupId>io.kroxylicious</groupId>
+            <artifactId>kroxylicious-kms-provider-kroxylicious-inmemory</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.kroxylicious</groupId>
+            <artifactId>kroxylicious-kms-test-support</artifactId>
+        </dependency>
```
Comment (SamBarker) [inline kroxylicious-kms-provider-kroxylicious-inmemory-test-support/pom.xml:32]: This module should depend on `junit-jupiter-api` and `assertj-core` as it directly uses them (Yes, I know they are available transitively from `kroxylicious-kms-test-support`).

Alternatively we could make `kroxylicious-kms-test-support` a parent module but I'm not sure about what that implies for third party implementors and our general maven usage pattern.

Comment (SamBarker) [reply kroxylicious-kms-provider-kroxylicious-inmemory-test-support/pom.xml:32]: We should do the direct dependency as part of this PR. anything wider ranging should happen separately.

File: kroxylicious-kms-provider-hashicorp-vault-test-support/pom.xml [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.kroxylicious</groupId>
+            <artifactId>kroxylicious-kms-test-support</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.kroxylicious</groupId>
+            <artifactId>kroxylicious-runtime</artifactId>
+        </dependency>
+
```
Comment (SamBarker) [inline kroxylicious-kms-provider-hashicorp-vault-test-support/pom.xml:37]: Similar to the in memory test-support module we should add dependencies on `junit-jupiter-api` and `assertj-core` here.

File: kroxylicious-kms-provider-kroxylicious-inmemory-test-support/src/test/java/io/kroxylicious/kms/provider/kroxylicious/inmemory/InMemoryTestKmsFacadeTest.java (hunk 15 above)
Comment (SamBarker) [reply kroxylicious-kms-provider-kroxylicious-inmemory-test-support/src/test/java/io/kroxylicious/kms/provider/kroxylicious/inmemory/InMemoryTestKmsFacadeTest.java:27]: Ah wait no, we already have assertions around that from the `AbstractTestKmsFacadeTest`.

File: kroxylicious-kms-test-support/src/main/java/io/kroxylicious/kms/service/AbstractTestKmsFacadeTest.java [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+        this.factory = factory;
+    }
+
+    @Test
+    void factory() {
+        try (var facade = factory.build()) {
+            assertThat(facade)
+                    .isNotNull()
+                    .extracting(TestKmsFacade::isAvailable)
+                    .isEqualTo(true);
```
Comment (SamBarker) [inline kroxylicious-kms-test-support/src/main/java/io/kroxylicious/kms/service/AbstractTestKmsFacadeTest.java:41]: Can we have this assertion? Doesn't this imply that we have to have `DockerClientFactory.instance().isDockerAvailable()` passing in order to run the build? 

I'm sure that holds for the project development team but feels a bit off as a requirement.

**Review by SamBarker — APPROVED** [review]
Awesome work @k-wall

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java (hunk 2 above)
Comment (robobario) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/VaultTemplateInvocationContext.java]: I prefer the guard clause approach, the throw at the start of the method is easier to read I find. Taking it further if you have multiple parameters to check I would much prefer that to all happen as a unit at the start of the method, rather than having requireNonNulls to read and parse scattered in the method.

**Review by robobario — APPROVED** [review]
LGTM

File: kroxylicious-kms-test-support/src/main/java/io/kroxylicious/kms/service/AbstractTestKmsFacadeTest.java (hunk 18 above)
Comment (k-wall) [reply kroxylicious-kms-test-support/src/main/java/io/kroxylicious/kms/service/AbstractTestKmsFacadeTest.java:41]: If Docker isn't available, io.kroxylicious.kms.provider.hashicorp.vault.VaultTestKmsFacadeTest#beforeEach prevents the test from even starting.   Is that too obscure?

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/EnvelopeEncryptionFilterIT.java (hunk 14 above)
Comment (k-wall) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/EnvelopeEncryptionFilterIT.java:266]: done.

File: kroxylicious-kms-test-support/src/main/java/io/kroxylicious/kms/service/AbstractTestKmsFacadeTest.java (hunk 18 above)
Comment (SamBarker) [reply kroxylicious-kms-test-support/src/main/java/io/kroxylicious/kms/service/AbstractTestKmsFacadeTest.java:41]: Ahh ok, no that works nicely :)

File: kroxylicious-kms-provider-kroxylicious-inmemory-test-support/pom.xml (hunk 16 above)
Comment (k-wall) [reply kroxylicious-kms-provider-kroxylicious-inmemory-test-support/pom.xml:32]: DOne.

File: kroxylicious-kms-provider-hashicorp-vault-test-support/pom.xml (hunk 17 above)
Comment (k-wall) [reply kroxylicious-kms-provider-hashicorp-vault-test-support/pom.xml:37]: Done.
